### DISCUSSION TITLE
aioble peripheral receives only 20 bytes

### DISCUSSION THREAD

**Discussion by iot49** [micropython/micropython#12398] [ESP32] [answered]
I use aioble to communicate between an ESP32 and WebBluetooth (Chrome).

The peripheral can send messages up to 509 bytes length, but receive only the first 20 bytes of messages sent by the central (the received message is simply clipped, without error from either side).

Setup:

```
aioble.config(mtu=512)
```

and once a connection is received

```
await connection.exchange_mtu(512)
```

The VM is a recent Micropython fetch from github. The central (Chrome) runs on a Mac M1.

I realize this could be an issue with Micropython or Chrome/Mac OS but am at a loss how to diagnose. Suggestions?

**Comment by andrewleech** [comment]
Oh, yeah this stuff is a massive pain to diagnose. I did a lot of this MTU work a few years ago, but on stm. 

A ble sniffer can be helpful to see what's going over the air (full size or truncated packets) but that's not always easy to get set up.

On the device side the problem could be in either the controller or the host stack; controller being supplied by esp, host by micropython.
I'm not sure if the esp controller supports buffering larger packets?
I'm also not sure if the default nimble configuration of micropython currently supports this either ... though I think it should.

**Comment by bixb922** [comment, marked as answer]
See also #11945, perhaps that can help

**Reply by iot49** [reply]
Thanks, got it working, although with a hack (last line below):

```
self._rx_characteristic = aioble.Characteristic(connector_service, _CONNECTOR_RX, indicate=True, write=True)
self._tx_characteristic = aioble.Characteristic(connector_service, _CONNECTOR_TX, indicate=True, read=True)
aioble.register_services(connector_service)
aioble.core.ble.gatts_set_buffer(self._rx_characteristic._value_handle, 512)
```

No success with `BufferedCharacteristic` (it has the wrong flags). And I could not figure out how to

```
the easiest is just to write a larger value to the characteristic when your server starts, which will grow the internal storage for that characteristic.
```

as suggested in #11945.

Unless I am missing something, `gatts_set_buffer` should perhaps be accessible from `aioble`, maybe in the form of a max_len flag for `server.Characteristic`.

**Reply by jimmo** [reply]
> And I could not figure out how to

You just call .write on the characteristic with a dummy value that's as big as the biggest value that the client might write to it.

> Unless I am missing something, gatts_set_buffer should perhaps be accessible from aioble, maybe in the form of a max_len flag for server.Characteristic.

@iot49 Yes, see https://github.com/micropython/micropython-lib/pull/629 (I will try and get this merged this week).

**Comment by iot49** [comment]
To the extent that my "hack" works (`aioble.core.ble.gatts_set_buffer(self._rx_characteristic._value_handle, 512)`), this can be closed.

**Comment by IAmCoder** [comment]
`BufferedCharacteristic` worked for me with `max_len`:

```
self._rx_characteristic = aioble.BufferedCharacteristic(connector_service, _CONNECTOR_RX, indicate=True, write=True, max_len=512)
self._tx_characteristic = aioble.BufferedCharacteristic(connector_service, _CONNECTOR_TX, indicate=True, read=True, max_len=512)
```
